File: jvm/core/main/scala/com/webank/eggroll/core/deepspeed/job/meta/SubmitJobResponse.scala
package com.webank.eggroll.core.deepspeed.job.meta

import com.google.protobuf.ByteString
import com.webank.eggroll.core.meta.NetworkingModelPbMessageSerdes.{ErProcessorFromPbMessage, ErProcessorToPbMessage}
import com.webank.eggroll.core.meta.{Deepspeed, ErProcessor}

import scala.language.implicitConversions
import scala.collection.JavaConverters._

case class SubmitJobResponse(sessionId: String, processors: Array[ErProcessor])

object SubmitJobResponse {
  implicit def serialize(src: SubmitJobResponse): Array[Byte] = {
    Deepspeed.SubmitJobResponse.newBuilder()
      .setSessionId(src.sessionId)
      .addAllProcessors(src.processors.toList.map(_.toProto()).asJava)
      .build().toByteArray
  }

  implicit def deserialize(byteString: ByteString): SubmitJobResponse = {
    val proto = Deepspeed.SubmitJobResponse.parseFrom(byteString)
    SubmitJobResponse(sessionId = proto.getSessionId,
      processors = proto.getProcessorsList.asScala.map(_.fromProto()).toArray)
  }
}

File: jvm/core/src/main/java/org/fedai/eggroll/core/pojo/SubmitJobResponse.java
package org.fedai.eggroll.core.pojo;

import com.google.protobuf.InvalidProtocolBufferException;
import com.webank.eggroll.core.meta.Deepspeed;
import lombok.Data;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.List;
import java.util.stream.Collectors;

@Data
public class SubmitJobResponse implements RpcMessage {
    Logger log = LoggerFactory.getLogger(SubmitJobResponse.class);
    private String sessionId;
    private List<ErProcessor> processors;

    @Override
    public byte[] serialize() {
        Deepspeed.SubmitJobResponse.Builder builder = Deepspeed.SubmitJobResponse.newBuilder()
                .setSessionId(this.sessionId)
                .addAllProcessors(this.processors.stream().map(ErProcessor::toProto).collect(Collectors.toList()));
        return builder.build().toByteArray();
    }

    @Override
    public void deserialize(byte[] data) {
        try {
            Deepspeed.SubmitJobResponse proto = Deepspeed.SubmitJobResponse.parseFrom(data);
            this.processors = proto.getProcessorsList().stream().map(ErProcessor::fromProto).collect(Collectors.toList());
            this.sessionId = proto.getSessionId();
        } catch (InvalidProtocolBufferException e) {
            log.error("deserialize error : ", e);
        }
    }

}
